Fixes dissonance between `disconnect_expired_cert` vs `require_session_mfa`

ibeckermayer opened the pull request:
Adds ssh/tls cert extensions
```
// CertExtensionMFAVerifiedSessionExpires is the RFC3339 timestamp representing the hard
// deadline of the session on a certificates issued after an MFA check.
// See https://github.com/gravitational/teleport/issues/18544.
CertExtensionMFAVerifiedSessionExpires = "mfa-verified-session-expires"

// MFAVerifiedSessionExpiresASN1ExtensionOID is the RFC3339 timestamp representing the hard
// deadline of the session on a certificates issued after an MFA check.
// See https://github.com/gravitational/teleport/issues/18544.
MFAVerifiedSessionExpiresASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 2, 12}
```

These get set when short lived (<1m) single-use MFA certs are issued, and ultimately passed to the connection monitor as the `DisconnectExpiredCert` value, which fixes https://github.com/gravitational/teleport/issues/18544

### Manual testing

See instructions in [webapps README](https://github.com/gravitational/webapps/blob/1e133f2e4975dc91db6f2cf1c8a14dcd7ecdf770/README.md?plain=1#L253-L271) for local mfa development. My auth service looks like
```yaml
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  authentication:
    second_factor: optional #off
    require_session_mfa: yes
    webauthn:
      rp_id: proxy.127.0.0.1.nip.io
  disconnect_expired_cert: yes
```

To confirm the force-disconnection time on mfa sessions, I added the line

```go
fmt.Printf("Monitor force-disconnecting in %v\n", discTime)
```

below this one in `monitor.go`
https://github.com/gravitational/teleport/blob/b571efe990b1b2db089199f002f4d83075eee335/lib/srv/monitor.go#L181

Before these changes it would show ~56seconds, whereas now it shows a time closer to the cluster's default ttl (12h in my case).

**Behavior manually confirmed for**
- [x] SSH (@ibeckermayer)
- [x] Desktop Access (@ibeckermayer)
- [x] Kubernetes (@tigrato)
- [x] App Access (@mdwn): will be double checked once https://github.com/gravitational/teleport/pull/18607#issuecomment-1323967391 is addressed, we can proceed with merging this fix for now.
- [x] DB Access (@smallinsky)

Commit:
Adds MFAVerifiedSessionExpires to certificates to prevent MFA sessions from being terminated after ~1min when disconnect_expired_cert is set

Commit:
Adds MFAVerifiedSessionExpires to the proto file and fills in the rest of the values, tested to work

Commit:
Adds MFAVerifiedSessionExpires checks to TestGenerateUserSingleUseCert

Commit:
Pushes DisconnectExpiredCert setting logic into functions and adds tests for on of the functions (the other is logically identical)

ibeckermayer commented on lib/kube/proxy/forwarder.go:
```diff
@@ -691,8 +691,14 @@ func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUs
 	disconnectExpiredCert := roles.AdjustDisconnectExpiredCert(authPref.GetDisconnectExpiredCert())
-	if !certExpires.IsZero() && disconnectExpiredCert {
-		authCtx.disconnectExpiredCert = certExpires
+	if disconnectExpiredCert {
+		if !identity.MFAVerifiedSessionExpires.IsZero() {
+			// Cause MFA verified sessions to disconnect on issuing certs expiry
+			// (see https://github.com/gravitational/teleport/issues/18544).
+			authCtx.disconnectExpiredCert = identity.MFAVerifiedSessionExpires
+		} else if !certExpires.IsZero() {
+			authCtx.disconnectExpiredCert = certExpires
+		}
```
I need some guidance here. I suspect this logic is wrong however I know next to nothing about how kubernetes access works in general or what this section is doing in particular.

The typical process for setting `disconnectExpiredCert` is encoded in `GetDisconnectExpiredCertFromIdentity`
https://github.com/gravitational/teleport/blob/6263b81de3ea284c6b93746a7c56a0ae954496f2/lib/srv/monitor.go#L429-L457

However because we are using a special value `certExpires` here, rather than the typical case of using `identity.Expires`, I suspect that using `identity.MFAVerifiedSessionExpires` on line 698 is wrong. Probably I instead want to parse the cert that `certExpires` come from and use the new extension (see top level description) that corresponds to `MFAVerifiedSessionExpires`, but I need somebody who actually understands the system to confirm.

Tagging those who `git blame` tells me have worked on code around here: @dboslee @tigrato @Joerger @smallinsky @klizhentas

ibeckermayer replied:
Here's what I think it should actually look like, still please somebody double check that this makes sense in context:

919aca472825265ea285559e2ec2fe544c4564aa

tigrato replied:
Kubernetes Proxy caches users' certificates and uses the `certExpires` (previously not after) to differentiate certificates for the same user (they will end up with different keys in the cache). This is important for cases where the user has connected to a proxy with a certificate that missed some Traits and has to re-login to regain them (check #18070 for more details)

As far as I can see, `Expires` is not marshaled into the certificate and it's not carried around. It depends on what is passed as the second arg to `tlsca.FromSubject` function.

```go
	id, err := tlsca.FromSubject(csr.Subject, time.Time{})
// or
        id, err := tlsca.FromSubject(csr.Subject,csr.NotAfter)
```
 and we have cases where we pass a zero time [[1](https://github.com/gravitational/teleport/blob/e17bcb13014e7b71bf787b3fcab5e8bd754d3b95/lib/auth/db.go#L152) ,[2](https://github.com/gravitational/teleport/blob/e17bcb13014e7b71bf787b3fcab5e8bd754d3b95/lib/auth/kube.go#L99)].

Since you pass the `clientCert.NotAfter` as a second arg to `tlsca.FromSubject`, `Expires` will be populated correctly.

When disconnect on cert expiration is enabled, the `disconnectExpiredCert` should be the cert expiration date if MFA is disabled or the `identity.MFAVerifiedSessionExpires` when MFA is enabled. 

From what I saw, your changes are correct and working accordingly.

ibeckermayer replied:
Thanks for testing

Commit:
derives a new identity in the kube forwarder to pass to GetDisconnectExpiredCertFromIdentity

zmb3 commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3211,6 +3211,13 @@ message Identity {
   ];
+  // MFAVerifiedSessionExpires is the the hard deadline of a session when this Identity was
+  // confirmed immediately after an MFA check.
+  google.protobuf.Timestamp MFAVerifiedSessionExpires = 23 [
```
I think we can do a better job with the name. Remember, the root issue is that the "disconnect expired cert" option doesn't work with session MFA. This is not really a session expiry, because there are valid use cases where a cert expires but the session is allowed to continue.

How about

ibeckermayer replied:
(Looks like you forgot to include your suggestion)

Yeah good point. How about `IssuingCertExpires`?

ibeckermayer replied:
73889219589f717c33193070365f12b631ae64c3

zmb3 commented on lib/auth/grpcserver.go:
```diff
@@ -2356,7 +2356,9 @@ func userSingleUseCertsAuthChallenge(gctx *grpcContext, stream proto.AuthService
-func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice) (*proto.SingleUseUserCert, error) {
+// mfaVerifiedSessionExpires is the expiry of the user cert from which this single use cert is derived from.
+// See https://github.com/gravitational/teleport/issues/18544.
+func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice, mfaVerifiedSessionExpires time.Time) (*proto.SingleUseUserCert, error) {
```
```suggestion
func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice) (*proto.SingleUseUserCert, error) {
```

You don't need to pass in the expiry as an argument, you already have it in `actx`.

zmb3 commented on lib/auth/grpcserver.go:
```diff
@@ -2356,7 +2356,9 @@ func userSingleUseCertsAuthChallenge(gctx *grpcContext, stream proto.AuthService
 }
 
-func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice) (*proto.SingleUseUserCert, error) {
+// mfaVerifiedSessionExpires is the expiry of the user cert from which this single use cert is derived from.
```
Is this comment misplaced?

ibeckermayer replied:
I think I was trying to document what the `previousIdentityExpires` parameter does, but that's documented enough elsewhere that I'll just get rid of it.

zmb3 commented on lib/auth/permissions.go:
```diff
@@ -297,18 +302,19 @@ func (a *authorizer) authorizeRemoteUser(ctx context.Context, u RemoteUser) (*Co
+		RouteToApp:                u.Identity.RouteToApp,
+		RouteToDatabase:           u.Identity.RouteToDatabase,
+		MFAVerified:               u.Identity.MFAVerified,
+		MFAVerifiedSessionExpires: mfaVerifiedSessionExpires,
```
This is either going to be empty, or identical to `Expires` right?

If that's the case, do we need to set it here or can we rely on the existing `Expires` and `MFAVerified` fields?

ibeckermayer replied:
@zmb3 I'm not sure what I was thinking, I think this makes sense: b571efe990b1b2db089199f002f4d83075eee335

@tigrato I'm not seeing how that relates

ibeckermayer replied:
> From what I understood, @zmb3 was asking to remove PreviousIdentityExpires here since its value is either empty or equal to Expires, and re-use the value of Expires and MFAVerified to decide the PreviousIdentityExpires.

I think what was meant was: at the time @zmb3's original comment was written, [the logic for setting `MFAVerifiedSessionExpires`](https://github.com/gravitational/teleport/pull/18607/files/919aca472825265ea285559e2ec2fe544c4564aa#diff-a0d9859c8f6c722d6937c51e77f3db91423e15cfebc385111d891ef30cdf218fR292-R296) (now called `PreviousIdentityExpires`) meant that it would either be empty or just the value of `Expires`, so it didn't really need to be set here.

However I don't know what I was thinking when I wrote that. My understanding is that what we're doing here is converting a local cluster identity to an identity that original identity is allowed to take up on a remote cluster. 

What I'm doing between https://github.com/gravitational/teleport/commit/b571efe990b1b2db089199f002f4d83075eee335 + https://github.com/gravitational/teleport/commit/a26f201546be180648fceee39c15da2abe427c49 is copying the original local cluster identity's `PreviousIdentityExpires` trait over, while ensuring it doesn't exceed the max allowed TTL for this new remote cluster identity by passing it through `AdjustSessionTTL` (the same process that was already being done for `Expires`).

zmb3 commented on lib/services/authority.go:
```diff
@@ -287,6 +287,9 @@ type UserCertParams struct {
 	// MFAVerified is the UUID of an MFA device when this Identity was
 	// confirmed immediately after an MFA check.
 	MFAVerified string
+	// MFAVerified is the the hard deadline of a session when this Identity was
```
Same comment as above. It's not a hard deadline, because if "disconnect expired cert" is disabled then it does absolutely nothing.

How about something like `OriginalIdentityExpires`?

ibeckermayer replied:
https://github.com/gravitational/teleport/pull/18607#discussion_r1028409180

zmb3 commented on lib/srv/monitor.go:
```diff
@@ -424,3 +425,61 @@ func (t *TrackingReadConn) UpdateClientActivity() {
+
+	// If the session doesn't need to be disconnected on cert expiry just return the default value.
+	if !checker.AdjustDisconnectExpiredCert(authPref.GetDisconnectExpiredCert()) {
+		var t time.Time
+		return t
```
```suggestion
		return time.Time{}
```

zmb3 reviewed (commented):
Overall change looks good.

I think the godocs/naming can be improved, and we might be able to reduce some of the plumbing and set the original expiry at the last possible moment before we issue the MFA certs, but I have Friday afternoon brain so I'll take a fresh look again next week.

Commit:
Changes naming for MFAVerifiedSessionExpires to PreviousIdentityExpires

Commit:
Uses GetDisconnectExpiredCertFromIdentity in place of inline logic in the app server

Commit:
Makes the logic of GetDisconnectExpiredCertFromIdentity more explicit by also checking for MFAVerified, so that if we end up setting PreviousIdentityExpires in non-mfa situations in the future, the application logic isn't broken

Commit:
removes unnecessary comment

Commit:
removes unnecessary parameter

Commit:
fixes previous identity logic

Commit:
fixes import order

Commit:
another attempt at fixing import order

tigrato commented on lib/auth/permissions.go:
```diff
@@ -273,6 +273,12 @@ func (a *authorizer) authorizeRemoteUser(ctx context.Context, u RemoteUser) (*Co
 	ttl = checker.AdjustSessionTTL(ttl)
+	var previousIdentityExpires time.Time
+	if u.Identity.MFAVerified != "" {
+		prevIdentityTTL := time.Until(u.Identity.Expires)
```
Isn't this the same as ttl?

ibeckermayer replied:
Yes, it was, changed to what I meant it to be: a26f201546be180648fceee39c15da2abe427c49

mdwn commented:
So I'm not sure that app access actually works with MFA enabled at the moment. I'm trying to figure out why. That being said, maybe not worth holding up on that verification.

Commit:
fixing prevIdentityTTL

Commit:
moves PreviousIdentityExpires up to the mapped identity section, hopefully fixes gci problem

Commit:
finally discovering the make command that takes care of everything for me

tigrato commented on lib/services/authority.go:
```diff
@@ -287,6 +287,9 @@ type UserCertParams struct {
 	// MFAVerified is the UUID of an MFA device when this Identity was
 	// confirmed immediately after an MFA check.
 	MFAVerified string
+	// MFAVerified is the the hard deadline of a session when this Identity was
```
Comment typo

tigrato commented on lib/srv/monitor.go:
```diff
@@ -424,3 +426,60 @@ func (t *TrackingReadConn) UpdateClientActivity() {
+		return time.Time{}
+	}
+
+	if identity.MFAVerified != "" && !identity.PreviousIdentityExpires.IsZero() {
```
Since PreviousIdentityExpires is only set when MFA is required, can't we just use 

```suggestion
	if !identity.PreviousIdentityExpires.IsZero() {
```

tigrato commented on lib/srv/monitor.go:
```diff
@@ -424,3 +426,60 @@ func (t *TrackingReadConn) UpdateClientActivity() {
+	//
+	// See https://github.com/gravitational/teleport/issues/18544
+
+	// If the session doesn't need to be disconnected on cert expiry just return the default value.
```
it's probably better to reuse the same code in both functions

Commit:
small CR fixes and adds a test to confirm MFAVerified and PreviousIdentityExpires are encoded into SSH certificates

ibeckermayer commented:
Noticed we didn't have any sanity check test of this kind so I added it here https://github.com/gravitational/teleport/commit/faeed096b10553c4c7953388948cdbdb63507dd2#diff-3057c3148b41a175f31650e1b9033e2148353a0d4b43b79404d0fdfe0330f56aR176-R198

Commit:
Merge branch 'master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

zmb3 commented on constants.go:
```diff
@@ -439,6 +439,12 @@ const (
 	// CertExtensionMFAVerified is used to mark certificates issued after an MFA
 	// check.
 	CertExtensionMFAVerified = "mfa-verified"
+	// CertExtensionPreviousIdentityExpires is the RFC3339 timestamp representing the
+	// expiry time of the identity/cert that this identity/cert was derived from. It
+	// is used to determine a session's hard deadline in cases where both
+	// require_session_mfa and disconnect_expired_cert are enabled.
+	// See https://github.com/gravitational/teleport/issues/18544.
+	CertExtensionPreviousIdentityExpires = "prev-identity-expires"
```
```suggestion
	// CertExtensionPreviousIdentityExpires is an extension that stores the RFC3339 timestamp representing the
	// expiry time of the identity/cert that this identity/cert was derived from. It
	// is used to determine a session's hard deadline in cases where both
	// require_session_mfa and disconnect_expired_cert are enabled.
	// See https://github.com/gravitational/teleport/issues/18544.
	CertExtensionPreviousIdentityExpires = "prev-identity-expires"
```

This variable itself is not a timestamp, so I would reword things a bit.

zmb3 commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3211,6 +3211,15 @@ message Identity {
     (gogoproto.nullable) = false
   ];
+  // PreviousIdentityExpires is the expiry time of the identity/cert that this
+  // identity/cert was derived from. It is used to determine a session's hard
```
Again, I wouldn't use the `hard deadline` verbiage, since the default mode is for disconnect on expired cert to be false (which means it isn't actually a "hard deadline" unless you opt-in to that behavior)

ibeckermayer replied:
I think it's accurate now, it says "It is used to determine a session's hard deadline **in cases where both require_session_mfa and disconnect_expired_cert are enabled**."

zmb3 commented on lib/auth/grpcserver.go:
```diff
@@ -2372,7 +2372,8 @@ func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req prot
 
 	// Generate the cert.
-	certs, err := actx.generateUserCerts(ctx, req, certRequestMFAVerified(mfaDev.Id), certRequestClientIP(clientIP))
+	previousIdentityExpires := actx.Identity.GetIdentity().Expires
+	certs, err := actx.generateUserCerts(ctx, req, certRequestMFAVerified(mfaDev.Id), certRequestPreviousIdentityExpires(previousIdentityExpires), certRequestClientIP(clientIP))
```
This is long enough I'd probably wrap each of the `certRequestXYZ` options on its own line. And then inline the `previousIdentityExpires` variable.

Commit:
small CR fixes

zmb3 commented on lib/auth/test/suite.go:
```diff
@@ -172,6 +172,30 @@ func (s *AuthSuite) GenerateUserCert(t *testing.T) {
+	require.NoError(t, err)
+	parsedCert, err = sshutils.ParseCertificate(cert)
+	require.NoError(t, err)
+	devId, ok := parsedCert.Extensions[teleport.CertExtensionMFAVerified]
+	require.True(t, ok)
+	require.Equal(t, "mfa-device-id", devId)
```
Up to you, but:
```suggestion
	require.Contains(t, parsedCert.Extensions, teleport.CertExtensionMFAVerified)
	require.Equal(t, "mfa-device-id", parsedCert.Extensions[teleport.CertExtensionMFAVerified])
```

zmb3 commented on lib/kube/proxy/forwarder.go:
```diff
@@ -536,7 +541,7 @@ func (f *Forwarder) formatResponseError(rw http.ResponseWriter, respErr error) {
 }
 
-func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUser bool, certExpires time.Time) (*authContext, error) {
+func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUser bool, clientIdentity *tlsca.Identity) (*authContext, error) {
```
While here, can we use another var name for the `auth.Context`? I prefer to keep `ctx` reserved for `context.Context`.

zmb3 commented on lib/srv/monitor.go:
```diff
@@ -424,3 +425,59 @@ func (t *TrackingReadConn) UpdateClientActivity() {
+
+// GetDisconnectExpiredCertFromIdentity calculates the proper value for DisconnectExpiredCert
+// based on whether a connection is set to disconnect on cert expiry, and whether
+// the cert is a short lived (<1m) one issued for an MFA verified session.
```
I'd mention under what circumstances it returns a zero time.

Commit:
fix comment

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

Commit:
Update lib/auth/test/suite.go

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

Commit:
updates test

Commit:
reserving ctx for context.Context

Commit:
adds note on when GetDisconnectExpiredCertFromIdentity returns the default value

Commit:
Merge remote-tracking branch 'origin/master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

Commit:
Id to ID

Commit:
Merge remote-tracking branch 'origin/master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

Commit:
Merge branch 'master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

Commit:
Fixes the broken lib/kube/proxy/forwarder_test::TestAuthenticate
test by adding the required Subject values such that tlsca.FromSubject()
can succeed in lib/kube/proxy/forwarder.go::Forwarder.authenticate().

I still need an expert on this code to confirm that we can rely on Subject
always being filled out here in production.

ibeckermayer commented:
@smallinsky @Joerger since I'm not familiar with this part of the code base, I'd like one of you guys to confirm that something for me, which boils down to confirming that we were never relying on calling `func (f *Forwarder) authenticate(req *http.Request) (*authContext, error)` on a request with a cert that was missing either of the fields I added (to a test) in a07e91315f561e1463972631ddf1b78d37abd5af.

The broader context is:

Initially my changes managed to break one of the forwarder tests (https://github.com/gravitational/teleport/issues/18928). I tracked it down, and the reason was that I'd added [this line](https://github.com/gravitational/teleport/pull/18607/files#diff-cdfe0b66f7090f9a5e0759d6cce323f7dde1ee1f0f49402d6a7652cd91b69430R436) to `func (f *Forwarder) authenticate(req *http.Request) (*authContext, error)`, which calls 

https://github.com/gravitational/teleport/blob/a07e91315f561e1463972631ddf1b78d37abd5af/lib/tlsca/ca.go#L292-L302 which needs a `Username` and `Groups`. I fixed it by modifying the test like this: a07e91315f561e1463972631ddf1b78d37abd5af.

Commit:
Merge remote-tracking branch 'origin/master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

Commit:
Merge branch 'master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

smallinsky commented:
@ibeckermayer Regarding: 
https://github.com/gravitational/teleport/pull/18607#issuecomment-1339938629

Based on my limited knowledge about teleport k8 access adding the [tlsca.FromSubject](https://github.com/gravitational/teleport/pull/18607/files#diff-cdfe0b66f7090f9a5e0759d6cce323f7dde1ee1f0f49402d6a7652cd91b69430R436) seems to be reasonable. 

The `(f *Forwarder) authenticate` function works only for `auth.RemoteUser` and `auth.LocalUser` contexts 
where  `tlsca.FromSubject`  is already called to create teleport context by the HTTP middleware: 
https://github.com/gravitational/teleport/blob/8c1e7be02dfec643d9796eff7a90a2119ac7e891/lib/auth/middleware.go#L489

Commit:
Merge branch 'master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

Commit:
Merge branch 'master' into isaiah/fix-disconnect_expired_cert-vs-require_session_mfa

ibeckermayer commented:
v9 has diverged substantially from v11 in various ways that make backporting difficult. I'm going to just leave this as-is for now, and can preform the surgical v9 backport if that's required by a particular customer.
